### PR TITLE
(rcm-34) implement new polling API, add back partial client, support configIDs

### PR DESCRIPTION
**Pull request by arbll**
### What does this PR do?

Continuation of the remote config refactoring started in https://github.com/DataDog/datadog-agent/pull/10142.

This PR:
- Adds back the partial client with a less hacky implementation
- Replaces the old pushing API by a polling API. Since tracers can not easily support pushing, this change goal is to reduce complexity by only supporting a single polling API for both tracers and agents. trace-agent will only act as a proxy here.
- Adds support for configuration IDs allowing multiple configuration to be consumed.
- Moves deserialization under remote-config's ownership
- Refactor the collection of tracer metadata to allow all type of clients to submit their information and state

### Review tips

There are three main components to review:
1. The partial client under `remote/uptane/partial_client.go`
2. The new API under `remoteconfig.proto` and `remote/service/service.go`
3. The new client handling configID and update logic under `remote/client.go`

### CODE REVIEW

File: pkg/config/remote/client.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	agentgrpc "github.com/DataDog/datadog-agent/pkg/util/grpc"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// Client is a remote-configuration client to obtain configurations from the local API
+type Client struct {
+	sync.Mutex
```
Comment (mellon85) [inline pkg/config/remote/client.go:22]: why promoting the Mutex if we anyway are going to use it only internally to the client?

File: pkg/config/remote/configs_apm_sampling.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+type apmSamplingConfigs struct {
+	config *APMSamplingConfig
+}
+
+func newApmSamplingConfigs() *apmSamplingConfigs {
+	return &apmSamplingConfigs{}
+}
+
+func (c *apmSamplingConfigs) update(configFiles map[string]configFiles) (*APMSamplingUpdate, error) {
```
Comment (mellon85) [inline pkg/config/remote/configs_apm_sampling.go:28]: probably we should pass around the configuration IDs as UUIDs to ensure they match the format and have a packed representation

File: pkg/config/remote/data/file.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+func ParseFilePathMeta(path string) (PathMeta, error) {
+	matchedGroups := filePathRegexp.FindStringSubmatch(path)
+	if len(matchedGroups) != filePathRegexpGroups+1 {
+		return PathMeta{}, fmt.Errorf("config file path '%s' has wrong format", path)
+	}
+	rawType := matchedGroups[1]
+	configType := TypeUnknown
+	switch rawType {
+	case "datadog":
+		configType = TypeDatadog
```
Comment (mellon85) [inline pkg/config/remote/data/file.go:46]: if it's unknown shouldn't this return an error?

File: pkg/config/remote/service/clients.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	expireAt time.Time
+	pbClient *pbgo.Client
+}
+
+func (c *client) expired(clock clock.Clock) bool {
+	return clock.Now().After(c.expireAt)
+}
+
+type clients struct {
+	sync.Mutex
```
Comment (mellon85) [inline pkg/config/remote/service/clients.go:21]: as before, do we really want to expose the internal Lock/Unlock state?

File: pkg/config/remote/data/file.go (hunk 3 above)
Comment (arbll) [reply pkg/config/remote/data/file.go:46]: The idea is to ignore anything that is not `datadog/` hence the `unknown`

File: pkg/config/remote/client.go (hunk 1 above)
Comment (arbll) [reply pkg/config/remote/client.go:22]: not really, it's just a pattern I picked up to write short `c.Lock()` instead of `c.mutex.Lock()`. I agree it's not great, especially with public structs

File: pkg/config/remote/configs_apm_sampling.go (hunk 2 above)
Comment (arbll) [reply pkg/config/remote/configs_apm_sampling.go:28]: I don't have a supper string opinion but not enforcing them agent side sounds ok to me.

File: pkg/config/remote/service/clients.go (hunk 4 above)
Comment (arbll) [reply pkg/config/remote/service/clients.go:21]: same, I'll make the change 👍🏻

File: pkg/config/remote/uptane/partial_client.go [hunk 5]
```diff
@@ -6,21 +6,250 @@
...
+	}
+	targetsRole, hasRoleTargets := root.Roles["targets"]
+	if !hasRoleTargets {
+		log.Warn("root is missing a targets role")
+		return nil
+	}
+	role := &data.Role{Threshold: targetsRole.Threshold, KeyIDs: targetsRole.KeyIDs}
+	if err := db.AddRole("targets", role); err != nil {
+		log.Warnf("could not add targets role to db: %v", err)
+		return nil
```
Comment (mellon85) [inline pkg/config/remote/uptane/partial_client.go:132]: Won't this break a whole product configuration deployment?

Comment (arbll) [reply pkg/config/remote/uptane/partial_client.go]: I changed those to actual errors to make it easier to reason about

File: pkg/proto/datadog/remoteconfig/remoteconfig.proto [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  string name = 4;
+  string version = 5;
+  bool is_tracer = 6;
+  TracerClient tracer_client = 7;
+}
+
+message TracerClient {
+  string service = 1;
+  string env = 2;
+  string version = 3;
```
Comment (raphaelgavache) [inline pkg/proto/datadog/remoteconfig/remoteconfig.proto:75]: This misses the tracer tag RuntimeID used to identify a tracer (in config.proto). I guess you wanted to merge with client.ID. I think it's better to keep it an explicit tracer tag. Probably client.ID can be removed as dupe of client.Name

Other tags you'll want for tracking and debugging tracers:
- appVersion (expliciting which version)
- language
- tracerVersion

File: pkg/config/remote/uptane/client.go [hunk 7]
```diff
@@ -153,6 +158,25 @@ func (c *Client) TargetFile(path string) ([]byte, error) {
 	return buffer.Bytes(), nil
 }
 
+// TargetsMeta returns the current raw targets.json meta of this uptane client
+func (c *Client) TargetsMeta() ([]byte, error) {
+	c.Lock()
+	defer c.Unlock()
+	err := c.verify()
```
Comment (raphaelgavache) [inline pkg/config/remote/uptane/client.go:165]: Why call c.verify() in DirectorRoot, Targets, TargetMeta while the verification is done at Update time?

File: pkg/config/remote/client.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+		return nil, err
+	}
+	var configFiles configFiles
+	for targetPath, target := range targets {
+		targetPathMeta, err := data.ParseFilePathMeta(targetPath)
+		if err != nil {
+			return nil, err
+		}
+		if _, productEnabled := c.enabledProducts[targetPathMeta.Product]; productEnabled {
+			targetContent, err := c.partialClient.TargetFile(targetPath)
```
Comment (raphaelgavache) [inline pkg/config/remote/client.go:170]: I think it would be clearer if the partial client returned a list of verified configs. All errors should already be caught by partialClient.Update(response)

verifiedConfigs := c.partialClient.GetVerifiedConfigs() 
// trim products

**Review by raphaelgavache — COMMENTED** [review]
partial tuf client + internal remote/client look good to me, left couple nits

File: pkg/config/remote/uptane/client.go (hunk 7 above)
Comment (arbll) [reply pkg/config/remote/uptane/client.go:165]: The idea is to make reasoning about security straightforward if those methods somehow get called after a failed update. This is a failsafe to ensure nothing can get extracted of the client if the local state is not valid.

**Review by KSerrania — APPROVED** [review]
LGTM for Agent Platform-owned files

File: pkg/config/remote/uptane/partial_client.go [hunk 9]
```diff
@@ -6,21 +6,247 @@
...
+		}
+	}
+	if targetFile == nil {
+		return nil, fmt.Errorf("target file %s not found", path)
+	}
+	targetMeta, hasMeta := c.targetMetas[path]
+	if !hasMeta {
+		return nil, fmt.Errorf("target file meta %s not found", path)
+	}
+	if len(targetMeta.HashAlgorithms()) == 0 {
```
[the remark this answers is not in the document]
Comment (arbll) [reply pkg/config/remote/uptane/partial_client.go:230]: Oh very nice 👍🏻

File: pkg/proto/datadog/remoteconfig/remoteconfig.proto (hunk 6 above)
Comment (arbll) [reply pkg/proto/datadog/remoteconfig/remoteconfig.proto]: Added the suggested ones but I can't remove client.ID as it's used to dedupe clients in the active client logic. Technically I could write an ID() method but this sounds error prone

**Review by raphaelgavache — APPROVED** [review]
lgtm!

File: pkg/proto/datadog/remoteconfig/remoteconfig.proto [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  string name = 4;
+  string version = 5;
+  bool is_tracer = 6;
+  TracerClient tracer_client = 7;
+}
+
+message TracerClient {
+  string runtime_id = 1;
+  string language = 2;
+  string tracer_version = 3;
```
Comment (ojung) [inline pkg/proto/datadog/remoteconfig/remoteconfig.proto:75]: I was thinking, since there might be more additional metadata that we might want to add in the future: should we have a `repeated string tags = x;` instead of adding new top-level fields to the message?

**Review by mellon85 — APPROVED** [review]

File: pkg/proto/datadog/remoteconfig/remoteconfig.proto (hunk 10 above)
Comment (arbll) [reply pkg/proto/datadog/remoteconfig/remoteconfig.proto:75]: The ones we currently have describe a tracer entirely and they make sense as top-level fields IMO. Adding a list of tag is fine to be future proof for optional meta we want to forward in the future but I don't have a strong opinion on wether or not this should be done right now. Since you're the stakeholder here I'll let you make the call :)

Comment (ojung) [reply pkg/proto/datadog/remoteconfig/remoteconfig.proto:75]: Nah, I suppose it doesn't have to be now. We can add it when we need it. 👍

File: pkg/trace/api/endpoints.go [hunk 11]
```diff
@@ -113,8 +113,8 @@ var endpoints = []endpoint{
 		Handler: func(r *HTTPReceiver) http.Handler { return r.debuggerProxyHandler() },
 	},
 	{
-		Pattern:   "/v0.6/config",
-		Handler:   func(r *HTTPReceiver) http.Handler { return http.HandlerFunc(r.handleConfig) },
+		Pattern:   "/v0.6/get_config",
```
Comment (gbbr) [inline pkg/trace/api/endpoints.go:116]: Why `get_`? Lets drop that please, unless there is a strong conventional reason.

File: pkg/trace/api/api.go [hunk 12]
```diff
@@ -71,13 +72,13 @@ type HTTPReceiver struct {
...
-	statsProcessor   StatsProcessor
-	appsecHandler    http.Handler
-	configSubscriber *config.Subscriber
+	out            chan *Payload
+	conf           *config.AgentConfig
+	dynConf        *sampler.DynamicConfig
+	server         *http.Server
+	statsProcessor StatsProcessor
+	appsecHandler  http.Handler
+	grpc           pbgo.AgentSecureClient
```
Comment (gbbr) [inline pkg/trace/api/api.go:81]:
```suggestion
	coreClient           pbgo.AgentSecureClient // gRPC client to core agent process
```
Is this correct? If yes, please rename it.

File: pkg/trace/api/endpoints.go (hunk 11 above)
Comment (arbll) [reply pkg/trace/api/endpoints.go]: Dropped and moved to v0.7

File: pkg/trace/api/api.go (hunk 12 above)
Comment (arbll) [reply pkg/trace/api/api.go]: done

File: pkg/trace/api/endpoints.go [hunk 13]
```diff
@@ -113,8 +113,8 @@ var endpoints = []endpoint{
 		Handler: func(r *HTTPReceiver) http.Handler { return r.debuggerProxyHandler() },
 	},
 	{
-		Pattern:   "/v0.6/config",
-		Handler:   func(r *HTTPReceiver) http.Handler { return http.HandlerFunc(r.handleConfig) },
+		Pattern:   "/v0.7/config",
```
Comment (gbbr) [inline pkg/trace/api/endpoints.go:116]: Just two commits ago this was `v0.6`. Why did you change it now?

Comment (arbll) [reply pkg/trace/api/endpoints.go:116]: It's breaking (we change the payload format), while we don't have official tracers supporting the older API for now there is some internal stuff using it and your comment reminded me of that

Comment (arbll) [reply pkg/trace/api/endpoints.go:116]: It's not necessary, just a nice to have

**Review by gbbr — APPROVED** [review]

File: pkg/trace/sampler/remote_rates.go [hunk 14]
```diff
@@ -40,40 +40,34 @@ type RemoteRates struct {
...
 
 func newRemoteRates(maxTPS float64) *RemoteRates {
 	if !features.Has("remote_rates") {
 		return nil
 	}
-	remoteRates := &RemoteRates{
+	client, err := remote.NewClient(remote.Facts{ID: "trace-agent", Name: "trace-agent", Version: version.AgentVersion}, []data.Product{data.ProductAPMSampling})
+	if err != nil {
+		log.Errorf("Error when subscribing to remote config management %v", err)
+		return nil
```
Comment (remeh) [inline pkg/trace/sampler/remote_rates.go:55]: returning `nil` on purpose I guess?

**Review by remeh — APPROVED** [review]
Looks good for files impacting agent-core

File: pkg/trace/sampler/remote_rates.go (hunk 14 above)
Comment (arbll) [reply pkg/trace/sampler/remote_rates.go:55]: Checked with APM and they are ok with this. Since the err can't be bubbled up easily it's either this or a fatal
